software/contrib/harmonic_lfos.py

```diff
@@ -41,8 +42,27 @@ def __init__(self):
 
         # Set the digital input and button handlers
         din.handler(self.reset)
-        b1.handler(self.change_mode)
-        b2.handler(self.increment_selection)
+
+        @b1.handler_falling
+        def b1Pressed():
+            """Triggered when B1 is pressed"""
+            if (
+                ticks_diff(ticks_ms(), b1.last_pressed()) > 1000
+                and ticks_diff(ticks_ms(), b1.last_pressed()) < 5000
+            ):
+                """Long press: toggle waveform view mode"""
+                self.viewAllWaveforms = not self.viewAllWaveforms
+            elif ticks_diff(ticks_ms(), b1.last_pressed()) < 500:
+                """Short press: Change the mode that controls wave shape"""
+                self.modes[self.selected_lfo] = (self.modes[self.selected_lfo] + 1) % self.MODES_COUNT
+
+            self.save_state()
```

Comment:
`[Recommended]`

Further to the above, if function calls inside an interrupt handler are bad, file-access is even worse.  Instead of saving the state inside the interrupt, you should set a flag (e.g. `self.state_needs_saving = True`) and then inside the main loop execute something like

```
if self.state_needs_saving:
    self.state_needs_saving = False
    self.save_state()
```

This way the (very slow) file i/o is handled outside the interrupt.

Reply:
I’m now realizing that if we implement your suggestion here, the timeout inside of save_state() may not be necessary at all. In other words, if we save outside of the interrupt handler and only when the user makes a change, we won’t need to have the timer to avoid the file i/o

Reply:
I believe with moving the `save_state()` call out of the ISR, the timer check may actually be a detriment.  If the state was last saved e.g. 2s ago, but it's changed again by the interrupt, with the current implementation `state_needs_saving` will be set to `False`, but the state _won't_ actually get saved, because the last save happened too recently.

I think you should remove
```
if self.last_saved() < 5000:
    return
```
from the `save_state` implementation